app/mm-redux/actions/posts.ts

```diff
@@ -406,36 +418,50 @@ export function editPost(post: Post) {
 export function getUnreadPostData(unreadChan: ChannelUnread, state: GlobalState) {
     const member = getMyChannelMemberSelector(state, unreadChan.channel_id);
     const delta = member ? member.msg_count - unreadChan.msg_count : unreadChan.msg_count;
+    const deltaRoot = member ? member.msg_count_root - unreadChan.msg_count_root : unreadChan.msg_count_root;
 
     const data = {
         teamId: unreadChan.team_id,
         channelId: unreadChan.channel_id,
         msgCount: unreadChan.msg_count,
         mentionCount: unreadChan.mention_count,
+        msgCountRoot: unreadChan.msg_count_root,
+        mentionCountRoot: unreadChan.mention_count_root,
         lastViewedAt: unreadChan.last_viewed_at,
         deltaMsgs: delta,
+        deltaMsgsRoot: deltaRoot,
     };
 
     return data;
 }
 
-export function setUnreadPost(userId: string, postId: string) {
+export function setUnreadPost(userId: string, postId: string, location: string) {
     return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
         let state = getState();
         const post = Selectors.getPost(state, postId);
         let unreadChan;
 
+        const collapsedThreadsEnabled = isCollapsedThreadsEnabled(state);
+        const isThreadUnread = collapsedThreadsEnabled && location === THREAD;
+
         try {
             if (isCombinedUserActivityPost(postId)) {
                 return {};
             }
-            unreadChan = await Client4.markPostAsUnread(userId, postId);
-            dispatch({
-                type: ChannelTypes.ADD_MANUALLY_UNREAD,
-                data: {
-                    channelId: post.channel_id,
-                },
-            });
+            if (isThreadUnread) {
+                const currentTeamId = getCurrentTeamId(state);
+                const threadId = post.root_id || post.id;
```

Comment:
if there is no post.root_id it means that you are in fact marking the root post as unread, but how do you know it actually has more posts in that thread? shouldn't you check for that before deciding to  call `handleFollowChanged` and `updateThreadRead` instead of calling `markPostAsUnread`?

Reply:
this is following the spec and seems to be expected